SMT Switch API

thk123 opened the pull request:
Adds the [smt-switch](https://github.com/makaimann/smt-switch) dependency if you build with `cmake and -DSMT_SWITCH=ON -DSMT_CVC4=ON`. Currently not used, outside of a unit test to prove compiling correctly. 

There are still a few things I need to do
 - [x] ~Need to add to `make` build as well I suppose.~
 - [x] Update documentation for how to enable the build
 - [x] Enable the codebuild job

But at least the CMake changes can be reviewed 

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Integrate smt-switch to the cbmc build process

Commit:
Add simple unit test for checking solver linking

The tests must exist, even if the smt-switch library is turned off as CI counts
the number of tests written is the same as the number of tests executed.

Commit:
Add a defintion for when compiling with SMT

Commit:
Install the CVC4 dependency if enabled

This is required to be able to link against the target as it installs the
necessary depenedices.

Smt-switch is a C++17 project so the target must be adjusted accordingly.

thk123 commented on CMakeLists.txt:
```diff
@@ -51,8 +52,22 @@ if(${enable_cbmc_tests})
+  download_project(
+    PROJ smt-switch
+    GIT_REPOSITORY  https://github.com/thk123/smt-switch.git
+    GIT_TAG patch-1
```
This resolves a `-pedantic` error that is currently got by downloading the master or last release version. I have a PR to fix this in trunk: makaimann/smt-switch#43

If that doesn't get merged soon, the alternative would be to disable `-pedantic` for this project (hopefully @hannes-steffenhagen-diffblue knows a way to do this)

thk123 commented on CMakeLists.txt:
```diff
@@ -51,8 +52,22 @@ if(${enable_cbmc_tests})
+  add_compile_options(-DHAVE_SMT_SWITCH)
+#  if(BUILD_CVC4)
+    execute_process(COMMAND ./contrib/setup-cvc4.sh WORKING_DIRECTORY ${smt-switch_SOURCE_DIR})
+#  endif()
```
I will add sections for each of the supported modules here, but might be worth having a think about whether we want one on by default, if you use smt-switch).

thk123 commented on unit/solvers/smt_switch.cpp:
```diff
@@ -0,0 +1,38 @@
+
+#include <vector>
+
+TEST_CASE("Trivial SMT switch integration", "[core]")
```
:no_entry_sign: not appropriately tagged

Commit:
Don't apply C++ propreties to java models library

Commit:
Only apply warnings flags for dependencies

Commit:
Add an codebuild job for testing smt switch

thk123 commented:
Can't add make since smt-switch can't be built with make anyway.

Commit:
Add documentation for compiling and using smt-switch

hannes-steffenhagen-diffblue commented on CMakeLists.txt:
```diff
@@ -136,6 +163,12 @@ function(cprover_default_properties)
 
+    foreach (target IN LISTS ARGN)
+        message(${target})
+        target_compile_options(${target} PRIVATE -Wall -Wpedantic -Werror -Wno-deprecated-declarations -Wswitch-enum)
```
For other reviewers: This means we only set our warning/werror flags for targets that have `cprover_default_properties` called for them. We should do a check to make sure this is the case for all our own targets. Not setting werror stuff globally is so we can avoid having to patch dependencies just so we can avoid generating a warning.

hannes-steffenhagen-diffblue commented on unit/solvers/smt_switch.cpp:
```diff
@@ -0,0 +1,43 @@
+
+// only enable the tests if compiling with SMT switch
+#ifdef HAVE_SMT_SWITCH
+#  include <cvc4_factory.h>
```
What’s with the spaces after `#`?

thk123 replied:
This is how my clang-format wants to format this :man_shrugging:

Commit:
Update documentation to include dependencies

The solvers have additional dependencies that must be installed before
running the cmake configuration step.

thk123 commented:
@peterschrammel I'm not sure if we want to merge this before it has a user. 

Advantages:
 - ensures it is tested by CI so not inadvertently broken

Downsides:
 - Code whose purpose isn't clear 

I guess if we do merge it, we should remove the travis job since that is the bottleneck on CI and it is tested on AWS codebuild anyway.

Commit:
Adapt unit test to be an xfail

This way can ensure that the test is being run on CI.

thk123 commented:
Fixed clang-format - I believe this will now pass CI 

@peterschrammel expressed a preference for merging this, perhaps him and @hannes-steffenhagen-diffblue could you review this?

thk123 commented:
Have manually confirmed that the smt switch unit test passes in CI (does not appear in the xfail step), but is in the xfail for non-smt build.

NlightNFotis commented:
What's the status on this one? Is this still going in, or is this now going to be closed?

thk123 commented:
Given we're not pushing this forward I'll just close this PR for now.
